### PR TITLE
[RELEASE-0.16] Read images from environment in the upstream operator

### PR DESCRIPTION
**Pull request by markusthoemmes**
This is a stop-gap solution to fix our issues with image override raciness between the serverless-operator and the upstream operator. Instead of reading the registry from the KnativeServing/Eventing instance, this overrides the observed registry with values built from the environment, just like we do in the serverless-operator.

This should allow us to consistently and correctly override images in the upstream operator without races.

### CODE REVIEW

**Comment by markusthoemmes** [comment]
/hold

As this is under investigation.

File: pkg/reconciler/common/transformers.go [hunk 1]
```diff
@@ -25,6 +25,11 @@ import (
 // transformers that are common to all components.
 func transformers(ctx context.Context, obj v1alpha1.KComponent) []mf.Transformer {
 	logger := logging.FromContext(ctx)
+
+	// TODO: Openshift Hack: Take image registry from environment instead of CR
+	imagesFromEnviron(obj)
```
Comment (matzew) [inline pkg/reconciler/common/transformers.go]: that's something we will than "patch" to all of our releases ? 

If so, looks like we should get some "create release branch" script in place :-) 

we arelady updated OWNERS and add docker image etc to the upstream relerase branch

**Comment by aliok** [comment]
Just tried on OpenShift along with https://github.com/openshift-knative/serverless-operator/pull/478.

Works nice.

/lgtm
/approve

File: pkg/reconciler/common/transformers.go (hunk 1 above)
Comment (markusthoemmes) [reply pkg/reconciler/common/transformers.go]: ack, but separately please.

Comment (matzew) [reply pkg/reconciler/common/transformers.go]: sure - just raising the "concern" - that;s all. I am not saying do all of it now :-) I can help here too ;-)

File: pkg/reconciler/common/image_override.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	"os"
+	"strings"
+
+	"knative.dev/operator/pkg/apis/operator/v1alpha1"
+)
+
+const imagePrefix = "IMAGE_"
+
+// imagesFromEnviron overrides registry images
+func imagesFromEnviron(obj v1alpha1.KComponent) {
```
Comment (rhuss) [inline pkg/reconciler/common/image_override.go]: Maybe `Environment` is better here as we are all used to it. (see also the other function's name). Just found out the golang uese `os.Environ()` which, tbh, is totally new to me (and might be due to a special twist in the English language that I don't know)

We should either change it here or change to Environ below in the private's function name.

File: pkg/reconciler/common/image_override.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	if ks, ok := obj.(*v1alpha1.KnativeServing); ok {
+		if qpVal, ok := reg.Override["queue-proxy"]; ok {
+			configure(ks, "deployment", "queueSidecarImage", qpVal)
+		}
+	}
+}
+
+func imageMapFromEnvironment() map[string]string {
```
Comment (rhuss) [inline pkg/reconciler/common/image_override.go]: or rename this function

File: pkg/reconciler/common/image_override.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+			// "IMAGE_env_var=docker.io/foo3"
+			// "IMAGE_deployment__env_var=docker.io/foo4"
+			// to
+			// container: docker.io/foo
+			// deployment/container: docker.io/foo2
+			// env_var: docker.io/foo3
+			// deployment/env_var: docker.io/foo4
+			name := strings.TrimPrefix(pair[0], imagePrefix)
+			name = strings.Replace(name, "__", "/", 1)
+			if pair[1] != "" {
```
Comment (rhuss) [inline pkg/reconciler/common/image_override.go]: nit: you could do the check before extracting the name to save some unnecessary effort and immediately continue the loop . But a bit less nitty: Would it make sense to throw an error here as you _expect_ to have a value part for the environment ?

File: pkg/reconciler/common/image_override.go (hunk 3 above)
Comment (rhuss) [reply pkg/reconciler/common/image_override.go]: I think this function deserves a test

File: pkg/reconciler/common/image_override.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+				overrideMap[name] = pair[1]
+			}
+		}
+	}
+	return overrideMap
+}
+
+func configure(ks *v1alpha1.KnativeServing, cm, key, value string) bool {
```
Comment (rhuss) [inline pkg/reconciler/common/image_override.go]: why return a bool if ignored when called ?

File: pkg/reconciler/common/image_override.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	if ks.Spec.Config == nil {
+		ks.Spec.Config = map[string]map[string]string{}
+	}
+
+	old, found := ks.Spec.Config[cm][key]
+	if found && value == old {
```
Comment (rhuss) [inline pkg/reconciler/common/image_override.go]: Does it harm if you override the old value even when it is the same value ? If not (and looks like that this is not an issue), I would safe this check and just unconditionally override the value regardless whether it is different or the same as this makes the flow more readable.

File: pkg/reconciler/common/transformers.go (hunk 1 above)
Comment (rhuss) [reply pkg/reconciler/common/transformers.go]: can we name this function a bit more descriptive ? (nit, again). As it stands now it looks like a get function that ignores it return value. What's about "configureImagesFromEnviron()" ?

File: pkg/reconciler/common/image_override.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
+package common
+
+import (
+	"os"
+	"strings"
+
+	"knative.dev/operator/pkg/apis/operator/v1alpha1"
+)
+
+const imagePrefix = "IMAGE_"
```
Comment (rhuss) [inline pkg/reconciler/common/image_override.go:10]: Maybe the prefix is a bit to generic and there is a certain likelihood that there might be an accidental environment variable set that also start with `IMAGE_` ? Maybe sth like "SLO_IMAGE" might be less dangerous ? (SLO for "ServerLess-Operator")

**Review by rhuss — COMMENTED** [review]
Looks good to me with some nit-pick comments. The only thing I really would add is a unit test for the env var parsing (with edge cases), as this can be easily be done and will be helpful as long as this 'fix' is active.

Even if this fix is supposed to be only temporarily, a test is useful.

Because you never know how long the stopgap will stay ("Nichts hält länger als ein Provisorium" as we say, nothing lasts longer than a stopgap)

File: pkg/reconciler/common/image_override.go (hunk 7 above)
Comment (markusthoemmes) [reply pkg/reconciler/common/image_override.go:10]: Hmm, I'd rather not change this given it's the same throughout the serverless-operator too. I think we have enough "control" over the upstream to at least know if and when this prefix would collide.

File: pkg/reconciler/common/image_override.go (hunk 4 above)
Comment (markusthoemmes) [reply pkg/reconciler/common/image_override.go]: In general it might make sense, yes. Practically though, there is no mechanism to surface this error to the reconcile process currently, which means that the patch area of this would be vastly increased and causing us a lot more trouble.

Seeing as we have E2E checks in place that make sure that no `gcr.io` based image makes it into our deployment, I think we at least have a fairly good idea of whether or not we butchered the list so I'm leaning towards leaving the failure mode as-is.

Comment (rhuss) [reply pkg/reconciler/common/image_override.go]: ok, sounds reasonable.

File: pkg/reconciler/common/image_override.go (hunk 7 above)
Comment (rhuss) [reply pkg/reconciler/common/image_override.go:10]: ok

**Review by rhuss — APPROVED** [review]
Thanks !

/lgtm
